### PR TITLE
Add custom trainer for training with weightmaps

### LINKED ISSUE #7: Add custom trainer with support for training with weight maps
We want to add a trainer that supports:
* [x] Weighting the important aspects in the image.
* [x] Allowing training on partial annotations

### PR DESCRIPTION
**Pull request by silvandeleemput**
Closes #7 

This PR adds the experimental `nnUNetTrainerV2Weighted` trainer class. It also adds the following to the docs:

**WARNING this trainer is still very experimental, also the training times can increase significantly using this trainer, w.r.t. other trainers!**

This trainer class implements a simple training scheme with support for training with weightmaps (annotated weight values at the voxel/pixel level).

Note that for this trainer to work weightmaps have to be supplied next to the preprocessed training data. 
First you would prepare your data as normal and run all the planning and preprocessing steps. 
Once you have all your preprocessed files inside the `nnUNet_preprocessed_data`, you put the weightmaps next to the preprocessed files as `.npy` files here: `nnUNet_preprocessed_data/TaskXXX_DIRECTORY/nnUNetData_plans_v2.1_stage0`. 
So that you have the files similar to the following formats there:
```
├── hippocampus_001.npy
├── hippocampus_001.npz
├── hippocampus_001.pkl
├── hippocampus_001_weightmap.npy
├── hippocampus_002.npy
├── hippocampus_002.npz
├── hippocampus_002.pkl
└── hippocampus_002_weightmap.npy
```

To simplify the weightmap generation processes we have added a utility CLI that can be used to generate weightmaps in a similar fashion to the original [U-Net paper](https://arxiv.org/pdf/1505.04597.pdf):
`nnUNet_diag_generate_unet_weightmaps --help`

Some caveats with the weightmap training, which are good to know before usage:
* You can experience very long training times. The data pipeline is currently not optimized to handle the weightmaps and takes a lot of additional overhead.
* 2D training is currently not supported. If you need it let me know, it should be easy to add.
* Dice loss has been disabled by default during training, since it does not support training with weightmaps.
* The reported Dice values during online evaluation do not take weightmaps into account and should be ignored.

### CODE REVIEW

**Review by nlessmann — DISMISSED** [review]
Nice, looks good! Did you try a full training run?

File: diag.md [hunk 1]
```diff
@@ -2,25 +2,72 @@
...
+└── hippocampus_002_weightmap.npy
+```
+
+To simplify the weightmap generation processes we have added a utility CLI that can be used to generate weightmaps in a similar fashion to the original [U-Net paper](https://arxiv.org/pdf/1505.04597.pdf):
+`nnUNet_diag_generate_unet_weightmaps --help`
+
+Some caveats with the weightmap training, which are good to know before usage:
+* You can experience very long training times. The data pipeline is currently not optimized to handle the weightmaps and takes a lot of additional overhead.
+* 2D training is currently not supported. If you need it let me know, it should be easy to add.
+* Dice loss has been disabled by default during training, since it does not support training with weightmaps.
```
[the remark this answers is not in the document]
Comment (silvandeleemput) [reply diag.md:67]: It is just the implementation used here, although I wonder how you would compute a correct pixel-wise weighted Dice loss. Do you have a reference by chance?

File: nnunet/utilities/diag/generate_unet_weightmaps.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        seg = load_segmentation_data(input_file=input_file)
+        wmap = compute_unet_border_weightmaps(
+            seg_input=seg, classes=classes, sigma=sigma, w_0=w_0
+        )
+        wmap = np.expand_dims(wmap, axis=0)  # Add a dummy dimension to the data
+        print(f"Outputing data: {wmap.shape} to {output_file}")
+        np.save(str(output_file), wmap)
+
+
+def auto_detect_classes(input_dir: Path, matching_pattern: str) -> Tuple[int]:
```
[the remark this answers is not in the document]
Comment (silvandeleemput) [reply nnunet/utilities/diag/generate_unet_weightmaps.py:93]: That would probably be better yes.

**Comment by silvandeleemput** [comment]
> Nice, looks good! Did you try a full training run?

I tried a full training run and it works. However, it is very slow w.r.t. training without weightmaps.

File: nnunet/utilities/diag/generate_unet_weightmaps.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (silvandeleemput) [reply nnunet/utilities/diag/generate_unet_weightmaps.py]: Implemented.

File: diag.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (silvandeleemput) [reply diag.md]: Hmm, I can't tell from the code alone if it is correct or optimal. For now, I'll leave the Dice out. We could always subclass this class and add a new trainer class with a weighted dice loss as well. But I think that it'll be some work.

[the remark this answers is not in the document]
Comment (nlessmann) [reply diag.md]: Sounds good

**Review by nlessmann — APPROVED** [review]

**Pull request merged**